game/deck.py

```diff
@@ -1,34 +1,81 @@
+from collections import defaultdict
+from copy import deepcopy
 from random import shuffle
-from typing import List, Dict
+from typing import List, Dict, Counter
 
 from game.card import Card, Suit
 
 SUITS = [Suit.CLUBS, Suit.DIAMONDS, Suit.HEARTS, Suit.SPADES]
 
 
 class Deck:
-    def __init__(self):
-        self.cards_left: List[Card] = [Card(suit, number) for suit in SUITS for number in range(1, 14)]
-        self.shuffle()
+    def __init__(self, initial_cards: List[Card] = None):
+        if initial_cards is None:
+            self._cards_left: List[Card] = [Card(suit, number) for suit in SUITS for number in range(1, 14)]
+            self.shuffle()
 
-        self.cards_left_amount = len(self.cards_left)
-        self.suits_left: Dict[Suit, int] = {suit: 13 for suit in SUITS}
-        self.number_left: Dict[int, int] = {number: 4 for number in range(1, 14)}
+            self.cards_left_amount = len(self._cards_left)
+            self._suits_left: Dict[Suit, int] = {suit: 13 for suit in SUITS}
+            self._number_left: Dict[int, int] = {number: 4 for number in range(1, 14)}
+            self._decks: List["Deck"] = []
```

Comment:
just out of curiosity, if you are having each deck to hold 2 other decks, when their cards are below 12, they want create 2 extra deck themselves?
or am I missing something?

Reply:
yes, what you miss is that the deck creation occurs only using pop and only when we have 12 cards there at the pop-stage. any way - those cards will never have 13 cards neither will getting popped